File: tests/lib/fixtures.ts
// Fixture walkers for the acceptance harness. Each walker yields typed
// case descriptors that the per-category test files in
// `tests/cross_repo/*.ts` consume.

import { fromFileUrl, join, resolve } from "jsr:@std/path@1";

export const REPO_ROOT = resolve(
  fromFileUrl(new URL("../..", import.meta.url)),
);

export const FIXTURES_DIR = join(REPO_ROOT, "tests/fixtures");

export interface SourceFixture {
  name: string;
  dir: string;
}

export interface TargetFixture {
  name: string;
  caseName: string | null;
  dir: string;
}

export interface SkillFixture {
  skill: string;
  caseName: string;
  dir: string;
}

async function listDirs(root: string): Promise<string[]> {
  const out: string[] = [];
  try {
    for await (const entry of Deno.readDir(root)) {
      if (entry.isDirectory) out.push(entry.name);
    }
  } catch {
    // Optional root.
  }
  return out.sort();
}

async function exists(path: string): Promise<boolean> {
  try {
    await Deno.lstat(path);
    return true;
  } catch {
    return false;
  }
}

export async function walkSourceFixtures(): Promise<SourceFixture[]> {
  const root = join(FIXTURES_DIR, "sources");
  const out: SourceFixture[] = [];
  for (const name of await listDirs(root)) {
    out.push({ name, dir: join(root, name) });
  }
  return out;
}

export async function walkTargetFixtures(): Promise<TargetFixture[]> {
  const root = join(FIXTURES_DIR, "targets");
  const out: TargetFixture[] = [];
  for (const name of await listDirs(root)) {
    const targetDir = join(root, name);
    // Two layouts:
    //   adapters/targets/<name>/{input,expected}/        — single case
    //   adapters/targets/<name>/<case>/{input,expected}/ — multi case
    if (await exists(join(targetDir, "input"))) {
      out.push({ name, caseName: null, dir: targetDir });
      continue;
    }
    for (const caseName of await listDirs(targetDir)) {
      const caseDir = join(targetDir, caseName);
      if (await exists(join(caseDir, "input"))) {
        out.push({ name, caseName, dir: caseDir });
      }
    }
  }
  return out;
}

export async function walkSkillFixtures(skill: string): Promise<SkillFixture[]> {
  const root = join(FIXTURES_DIR, "skills", skill);
  const out: SkillFixture[] = [];
  for (const caseName of await listDirs(root)) {
    out.push({ skill, caseName, dir: join(root, caseName) });
  }
  return out;
}

File: tooling/src/test_support/fixtures.rs
use std::fs;
use std::path::{Path, PathBuf};

use super::fixtures_dir;
use crate::Context;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SourceFixture {
    pub name: String,
    pub dir: PathBuf,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TargetFixture {
    pub name: String,
    pub case_name: Option<String>,
    pub dir: PathBuf,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SkillFixture {
    pub skill: String,
    pub case_name: String,
    pub dir: PathBuf,
}

fn list_dirs(root: &Path) -> Vec<String> {
    let Ok(entries) = fs::read_dir(root) else {
        return Vec::new();
    };
    let mut names: Vec<String> = entries
        .filter_map(|entry| entry.ok())
        .filter(|entry| entry.file_type().is_ok_and(|ft| ft.is_dir()))
        .filter_map(|entry| entry.file_name().into_string().ok())
        .collect();
    names.sort();
    names
}

/// Every source fixture under `tests/fixtures/sources/<name>/`.
pub fn walk_source_fixtures(ctx: &Context) -> Vec<SourceFixture> {
    let root = fixtures_dir(ctx).join("sources");
    list_dirs(&root)
        .into_iter()
        .map(|name| {
            let dir = root.join(&name);
            SourceFixture { name, dir }
        })
        .collect()
}

/// Every target fixture under `tests/fixtures/targets/<name>/[<case>/]`.
pub fn walk_target_fixtures(ctx: &Context) -> Vec<TargetFixture> {
    let root = fixtures_dir(ctx).join("targets");
    let mut out = Vec::new();
    for name in list_dirs(&root) {
        let target_dir = root.join(&name);
        if target_dir.join("input").is_dir() {
            out.push(TargetFixture {
                name,
                case_name: None,
                dir: target_dir,
            });
            continue;
        }
        for case_name in list_dirs(&target_dir) {
            let case_dir = target_dir.join(&case_name);
            if case_dir.join("input").is_dir() {
                out.push(TargetFixture {
                    name: name.clone(),
                    case_name: Some(case_name),
                    dir: case_dir,
                });
            }
        }
    }
    out
}

/// Every skill fixture under `tests/fixtures/skills/<skill>/<case>/`.
pub fn walk_skill_fixtures(ctx: &Context, skill: &str) -> Vec<SkillFixture> {
    let root = fixtures_dir(ctx).join("skills").join(skill);
    list_dirs(&root)
        .into_iter()
        .map(|case_name| {
            let dir = root.join(&case_name);
            SkillFixture {
                skill: skill.to_string(),
                case_name,
                dir,
            }
        })
        .collect()
}
